### PR TITLE
Core & Internals: Added support for recursive lookup of attached DIDs when searching for DIDs; Fix #65

### LINKED ISSUE #65: rucio list-dids --recursive does not work
In the code it seems like this is not properly implemented.

_Migrated from JIRA [RUCIO-2794](https://its.cern.ch/jira/browse/RUCIO-2794)_

### PR DESCRIPTION
**Pull request by hahahannes**
Core & Internals: Added support for recursive lookup of attached DIDs when searching for DIDs; Fix #65

### CODE REVIEW

File: bin/rucio [hunk 1]
```diff
@@ -506,57 +506,53 @@ def list_dids(args):
...
-                    elif '=' in option:
-                        key, value = option.split('=')
-
-                    if key == 'type':
-                        if value.upper() in ['ALL', 'COLLECTION', 'CONTAINER', 'DATASET', 'FILE']:
-                            type = value.lower()
-                        else:
-                            logger.error('{0} is not a valid type. Valid types are {1}'.format(value, ['ALL', 'COLLECTION', 'CONTAINER', 'DATASET', 'FILE']))
-                            return FAILURE
-                    elif key == 'length.gt' or key == 'length.lt' or key == 'length.gte' or key == 'length.lte' or key == 'length':
```
Comment (vingar) [inline bin/rucio:544]: suggestion: if key in ('length.gt', etc)

File: bin/rucio [hunk 2]
```diff
@@ -506,57 +506,53 @@ def list_dids(args):
     type = 'all'
     table = []
     if args.filter:
-        if args.recursive:
-            logger.error('Option recursive and filter cannot be used together')
-            return FAILURE
-        else:
```
Comment (vingar) [inline bin/rucio:512]: else can be removed

File: bin/rucio [hunk 3]
```diff
@@ -506,57 +506,53 @@ def list_dids(args):
...
+                    if value.upper() in ['ALL', 'COLLECTION', 'CONTAINER', 'DATASET', 'FILE']:
+                        type = value.lower()
                     else:
-                        if value.lower() == 'true':
-                            value = '1'
-                        elif value.lower() == 'false':
-                            value = '0'
+                        logger.error('{0} is not a valid type. Valid types are {1}'.format(value, ['ALL', 'COLLECTION', 'CONTAINER', 'DATASET', 'FILE']))
+                        return FAILURE
+                elif key == 'length.gt' or key == 'length.lt' or key == 'length.gte' or key == 'length.lte' or key == 'length':
```
Comment (vingar) [inline bin/rucio]: suggestion: if key in ('length.gt', etc)

File: bin/rucio [hunk 4]
```diff
@@ -506,57 +506,53 @@ def list_dids(args):
...
+                        key = 'length.lte'
+                elif '<' in option:
+                    key, value = option.split('<')
+                    if key == 'length':
+                        key = 'length.lt'
+                elif '=' in option:
+                    key, value = option.split('=')
+
+                if key == 'type':
+                    if value.upper() in ['ALL', 'COLLECTION', 'CONTAINER', 'DATASET', 'FILE']:
```
Comment (mlassnig) [inline bin/rucio:535]: value.lower() so you save on the .lower() call next? :-D

Comment (mlassnig) [reply bin/rucio:535]: ah sorry, i'm blind, forget it

File: lib/rucio/tests/test_bin_rucio.py [hunk 5]
```diff
@@ -788,3 +788,47 @@ def test_detach_non_existing_file(self):
...
         print(err)
         nose.tools.assert_not_equal(re.search("Data identifier not found.", err), None)
+
+    def test_list_did_recursive(self):
+        """ CLIENT(USER): List did recursive """
+        # Setup nested collections
+        tmp_scope = 'mock'
+        tmp_container_1 = 'container_%s' % generate_uuid()
+        cmd = 'rucio add-container {0}:{1}'.format(tmp_scope, tmp_container_1)
+        print(self.marker + cmd)
```
Comment (mlassnig) [inline lib/rucio/tests/test_bin_rucio.py]: the prints in the testcase should be removed

**Review by mlassnig — APPROVED** [review]

**Pull request merged**